### PR TITLE
Initial sockmap implementation

### PR DESCRIPTION
**Pull request by jrfastab**
**Summary of changes**: This series adds initial sockmap support. The feature is initial behind a feature flag --sockops-enable. Currently, only IPv4 and conntrack timer updates are not supported. After initial support is added the remaining feature will be added.

The implementation relies on two new BPF programs (located in ./bpf/sockops/) one program (./bpf/sockops/bpf_sockops.c) monitors TCP events and when a connection enters the Established state and the destination socket is a local Cilium managed endpoint adds the socket to global sockmap. The global sockmap contains all Cilium managed sockets. A second program (./bpf/sockops/bpf_redir.c) is attached to the global sockmap and every msg sent over sockets in the map will run the BPF program. The BPF program will redirect the msg to the ingress queue of the peer socket. For example with L7 proxy enabled the redirect will be enabled on all traffic being sent through the L7 proxy.

This is an initial PR future PRs will address the following,
- IPv6
- events for errors
- conntrack updates
- currently every msg send has to do a policy lookup this is to handle cases where the policy is updated on existing sockets. This check can be removed if control plane can recalculate and remove sockets if necessary.
- CI testing infrastructure needs to support new kernels to test sockmap.

Before committing this PR:
- Rebase to latest Cilium
- BPF programs have dbg printing on... needs to be disabled
- Run ginkgo tests with sockops enabled.

**How to test**:

To test users will need a new kernel. Minimum kernel version is 4.17. And the option `--sockops-enable` must be used with cilium-agent.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5895)

### CODE REVIEW

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

File: bpf/lib/eps.h [hunk 1]
```diff
@@ -34,16 +34,35 @@ lookup_ip6_endpoint(struct ipv6hdr *ip6)
 	return map_lookup_elem(&cilium_lxc, &key);
 }
 
+static __always_inline struct endpoint_info *
+__lookup_ip4_endpoint(uint32_t ip)
+{
+	struct endpoint_key key= {};
```
Comment (vadorovsky) [inline bpf/lib/eps.h:40]: missing space?

File: bpf/sockops/bpf_sockops.c [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	 * pulled in as needed.
+	 */
+	svc = __lb4_lookup_service(&lb4_key);
+	if (svc) {
+		bpf_printk("ipv4 endpoint: services required use L4/L3 stack\n");
+		return;
+	}
+
+	/* Policy lookup required to learn proxy port */
+	if (1) {
```
Comment (vadorovsky) [inline bpf/sockops/bpf_sockops.c:62]: What's the purpose of checking whether 1 is true?

File: pkg/sockops/sockops.go [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+
+	// Mutex to sync access to sockops Root
+	sockopsMutex lock.Mutex
+
+	contextTimeout = 5 * time.Minute
+
+	once sync.Once
+
+	// BPF Program load path
+	cSockops = "bpf_sockops.c"
```
Comment (vadorovsky) [inline pkg/sockops/sockops.go:65]: const? Or are those variables supposed to be mutable?

File: pkg/sockops/sockops.go [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+		map_arg_list = append(map_arg_list, map_string...)
+	}
+
+	args := []string{"-m", "prog", "load", bpfObject, bpffs}
+	args = append(args, map_arg_list...)
+	log.WithFields(logrus.Fields{
+		"bpftool": prog,
+		"args":    args,
+	}).Debug("Load BPF Object:")
+	fmt.Printf("\nLoad BPF: %s %s\n", prog, args)
```
Comment (vadorovsky) [inline pkg/sockops/sockops.go:263]: log.Info? log.Debug?

Comment (jrfastab) [reply pkg/sockops/sockops.go:263]: debug code I'll just remove it.

File: pkg/sockops/sockops.go (hunk 3 above)
Comment (jrfastab) [reply pkg/sockops/sockops.go:65]: yeah lots of const floating in the code still probably can clean it up.

File: bpf/sockops/bpf_sockops.c (hunk 2 above)
Comment (jrfastab) [reply bpf/sockops/bpf_sockops.c:62]: none AFAIK just copy/paste from lxc code paths.

File: bpf/lib/eps.h (hunk 1 above)
Comment (jrfastab) [reply bpf/lib/eps.h:40]: yep.

**Comment by jrfastab** [comment]
closing this PR I have a new branch with updated code. Keeping this branch around however for reference. Comments still unaddressed are const fields and `if (1)` in bpf code. I think there is a reason for the `if (1)` it just escapes me at the moment.

**Pull request closed without merging**